qiskit/transpiler/passes/synthesis/high_level_synthesis.py

```diff
@@ -11,35 +11,130 @@
 # that they have been altered from the originals.
 
 
-"""Synthesize high-level objects."""
+"""Synthesize higher-level objects."""
+
 
 from qiskit.converters import circuit_to_dag
 from qiskit.transpiler.basepasses import TransformationPass
 from qiskit.dagcircuit.dagcircuit import DAGCircuit
-from qiskit.quantum_info.synthesis.clifford_decompose import decompose_clifford
+from qiskit.transpiler.exceptions import TranspilerError
+from .high_level_synthesis_plugins import HighLevelSynthesisPluginManager
+
+
+class HLSConfig:
+    """
+    For each higher-level-object (e.g., "clifford", "linear_function", etc.) the
+    high-level-synthesis config allows to specify a list of "methods", each method
+    represented by a pair consisting of a name of the synthesis algorithm and of a dictionary
+    providing additional arguments for this algorithm.
+
+    The names of the synthesis algorithms should be declared in ``entry_points`` for
+    ``qiskit.synthesis`` in ``setup.py``, in the form
+    <higher-level-object-name>.<synthesis-method-name>.
+
+    The standard higher-level-objects are recommended to have a synthesis method
+    called "default", which would be called automatically when synthesizing these objects,
+    without having to explicitly set these methods in the config.
+
+    To avoid synthesizing a given higher-level-object, one can give it an empty list of methods.
```

Comment:
Can you add a comment here on how the pass will handle cases where an `Operation` could be of multiple types when an `HLSConfig` is provided? e.g. A `LinearFunction` operation _could_ be synthesized by any of the available synthesis methods for a `Clifford`.

My expectation is that when you specify an `HLSConfig`, the list of applicable synthesis methods is strictly tied to the name of the type of the `Operation`, but it would be good to clarify this. (I also want to leave room for the case when a user doesn't specify an `HLSConfig`, for the pass to automatically select applicable synthesis methods which may include those from other types.)

Reply:
Right now the list of applicable synthesis methods is indeed strictly tried to the name of the `Operation` (I have tried to clarify this in the documentation following your comment). 

I completely agree that we want to have a way to synthesize `LinearFunctions` with synthesis methods for `Cliffords`, but at the moment I don't see a perfect solution for this. Internally, `LinearFunctions` are represented very differently from `Cliffords` (the former is a `nxn` binary linear matrix, the latter is represented using stabilizer/destabilizer tableaux -- or maybe this has changed recently), so we can't apply `Clifford` methods to `LinearFunctions` without first converting `LinearFunctions` into `Cliffords`. Do we want to have a transpiler pass for that? Or do we want to synthesize linear functions into CX-gates and then re-collect these as Clifford gates (with possibly additional adjacent `Clifford` gates)? And what should eventually be added to the optimization loop in terms of resynthesizing linear and clifford subcircuits?

Reply:
For now, it's sufficient to clarify the behavior of the current pass.

Moving forward, we'll likely need some investigation to weigh the relative cost of converting from the binary linear matrix we use for `LinearFunctions` to a stabilzer tableau and the likelihood of finding an improved synthesis by using one of the available `Clifford` methods over what is known for `LinearFunction`s. I could imagine the implementations being different depending on the answers there (and likewise for anywhere there is overlap of available synthesis methods, e.g. LinearFunctions  -> Clifford, Clifford -> Unitary, etc.)